theano/tensor/nnet/abstract_conv.py

```diff
@@ -1592,6 +1586,47 @@ def conv(self, img, kern, mode="valid", dilation=1, unshared=False):
             raise NotImplementedError('only 2D and 3D convolution are implemented')
         return out
 
+    def unshared2d(self, inp, kern, out_shape, direction=0):
+        '''
+        Basic slow Python unshared 2d convolution.
+        direction can be 0, 1 or 2 for forward pass,
+        gradWeights or gradInputs respectively
+        '''
+        if self.convdim != 2:
+            raise NotImplementedError('Unshared convolution not implemented for %dD'
+                                      % self.convdim)
+        if direction == 0:
+            if (kern.shape[0], kern.shape[1]) != (out_shape[0], out_shape[1]):
+                raise ValueError('Kernel shape ({},{}) does not match '
+                                 'output size ({},{})'.format(kern.shape[0], kern.shape[1],
+                                                              out_shape[0], out_shape[1]))
+        if direction not in (0, 1, 2):
+            raise ValueError("Direction must be one of 0, 1 or 2.")
+
+        out = np.zeros(out_shape, dtype=inp.dtype)
+
+        for row in xrange(out_shape[0]):
+            for col in xrange(out_shape[1]):
+                if direction == 0:
+                    out[row, col] += np.sum(np.multiply(inp[row:row + kern.shape[2],
+                                                        col:col + kern.shape[3]],
+                                            kern[row, col, ::-1, ::-1]))
+                elif direction == 1:
+                    out[row, col, ...] += kern[row, col] * \
+                        inp[row:row + out_shape[2], col:col + out_shape[3]]
+
+                else:  # direction = 2
+                    for k_row in xrange(kern.shape[2]):
+                        for k_col in xrange(kern.shape[3]):
+                            # Deciding which kernel to select
+                            reg_row = row - ((kern.shape[2] - 1) - k_row)
+                            reg_col = col - ((kern.shape[3] - 1) - k_col)
+                            if (reg_row >= 0 and reg_row < kern.shape[0]) and \
+                                    (reg_col >= 0 and reg_col < kern.shape[1]):
+                                out[row, col] += kern[reg_row, reg_col, k_row, k_col] * \
+                                    inp[reg_row, reg_col]
```

Comment:
I think this is more complicated than it needs to be. If you loop over the correct range, I don't think you need these `if`s checking whether you're inside the range.

Reversing your [forward convolution](https://github.com/vikramnitin9/Theano/blob/d803e2784a457f71a6eb31c88271c4ca15631d32/theano/tensor/nnet/abstract_conv.py#L1568-L1579), these loops give the same result for the grad wrt inputs:
```python
for row in xrange(kern.shape[0]):
    for col in xrange(kern.shape[1]):
        for k_row in xrange(kern.shape[2]):
            for k_col in xrange(kern.shape[3]):
                out[row + k_row, col + k_col] += inp[row, col] * \
                    kern[row, col, -(k_col + 1), -(k_row + 1)]
```